sa/sa.go

```diff
@@ -829,39 +857,88 @@ func (ssa *SQLStorageAuthority) RevokeCertificate(ctx context.Context, req *sapb
-	if req.Serial == "" || req.DateNS == 0 || req.BackdateNS == 0 {
+	if req.Serial == "" || req.DateNS == 0 || req.BackdateNS == 0 || req.IssuerID == 0 {
 		return nil, errIncompleteRequest
 	}
 	if req.Reason != ocsp.KeyCompromise {
 		return nil, fmt.Errorf("cannot update revocation for any reason other than keyCompromise (1); got: %d", req.Reason)
 	}
 
-	thisUpdate := time.Unix(0, req.DateNS)
-	revokedDate := time.Unix(0, req.BackdateNS)
+	_, overallError := db.WithTransaction(ctx, ssa.dbMap, func(tx db.Executor) (interface{}, error) {
+		thisUpdate := time.Unix(0, req.DateNS)
+		revokedDate := time.Unix(0, req.BackdateNS)
 
-	res, err := ssa.dbMap.ExecContext(ctx,
-		`UPDATE certificateStatus SET
-				revokedReason = ?,
-				ocspLastUpdated = ?
-			WHERE serial = ? AND status = ? AND revokedReason != ? AND revokedDate = ?`,
-		revocation.Reason(ocsp.KeyCompromise),
-		thisUpdate,
-		req.Serial,
-		string(core.OCSPStatusRevoked),
-		revocation.Reason(ocsp.KeyCompromise),
-		revokedDate,
-	)
-	if err != nil {
-		return nil, err
-	}
-	rows, err := res.RowsAffected()
-	if err != nil {
-		return nil, err
-	}
-	if rows == 0 {
-		// InternalServerError because we expected this certificate status to exist,
-		// to already be revoked for a different reason, and to have a matching date.
-		return nil, berrors.InternalServerError("no certificate with serial %s and revoked reason other than keyCompromise", req.Serial)
+		res, err := tx.ExecContext(ctx,
+			`UPDATE certificateStatus SET
+					revokedReason = ?,
+					ocspLastUpdated = ?
+				WHERE serial = ? AND status = ? AND revokedReason != ? AND revokedDate = ?`,
+			revocation.Reason(ocsp.KeyCompromise),
+			thisUpdate,
+			req.Serial,
+			string(core.OCSPStatusRevoked),
+			revocation.Reason(ocsp.KeyCompromise),
+			revokedDate,
+		)
+		if err != nil {
+			return nil, err
+		}
+		rows, err := res.RowsAffected()
+		if err != nil {
+			return nil, err
+		}
+		if rows == 0 {
+			// InternalServerError because we expected this certificate status to exist,
+			// to already be revoked for a different reason, and to have a matching date.
+			return nil, berrors.InternalServerError("no certificate with serial %s and revoked reason other than keyCompromise", req.Serial)
+		}
+
+		if req.ShardIdx != 0 {
+			var rcm revokedCertModel
+			// Note: this query MUST be updated to enforce the same preconditions as
+			// the UPDATE above, if this query ever becomes the first or only query
+			// in this transaction. We are currently relying on the query above to
+			// exit early if the certificate does not have an appropriate status.
+			err = tx.SelectOne(
+				ctx, &rcm, `SELECT * FROM revokedCertificates WHERE serial = ?`, req.Serial)
+			if err != nil && !db.IsNoRows(err) {
+				return nil, fmt.Errorf("retrieving revoked certificate row: %w", err)
+			} else if db.IsNoRows(err) {
```

Comment:
The nesting of this if/else/else is a little dissatisfying to me and per our conversation just now it sounds like it's a little dissatisfying to you too. One problem is the ordering: usually I prefer to handle the specific error case first and then the general error case:

```
if db.IsNoRows(err) {
  ...
  return ...
} else if err != nil {
  ...
  return ...
}
// success?
```

The other issue is that most of our error handling (not all!) follows the "early return" pattern, where errors usually beget other errors and are returned. In this case, the `IsNoRows` is actually expected to not "early return" in most cases (unless inserting or selecting a row runs into errors). So we can't follow the "fall through to success" model.

Also in general the nesting is a little deep and makes it a bit hard to spot that there are multiple "fall through to success" cases.

Some proposed fixes:

 - Handle the specific error case first (IsNoRows)
 - To mitigate the length of the IsNoRows case, factor it out into its own function, and document the preconditions explicitly. Perhaps the `SELECT FROM serials` / `Insert...revokedCertModel` couplet can be extracted to a function that is also used by `RevokeCertificate`?
 - Explicitly `return nil, nil` in each branch rather than falling through to success. `return nil, nil` is a single line of code and its meaning is clear; there's not need to avoid duplicating it.